infra/relayer/src/purse.js

```diff
@@ -0,0 +1,696 @@
+     * becomes available...  The order is also important here.  According to the
+     * HD wallet "standards" the lowest index/path should be uesd first.  That
+     * way, any wallets using the mnemonic in the future will discover all the
+     * children and their funds.  It is supposed to stop at the first one
+     * without a tx history.
+     */
+    do {
+      let lowestPending = this.maxPendingPerAccount
+      for (const child of this.children) {
+        const childBal = stringToBN(await this.web3.eth.getBalance(child))
+        if (
+          this.accounts[child].pendingCount < lowestPending &&
+          childBal.gt(MIN_CHILD_BALANCE)
+        ) {
+          lowestPending = this.accounts[child].pendingCount
+          resolvedAccount = child
+          logger.debug(`selecting account ${child}`)
+        }
+      }
+
+      if (resolvedAccount) {
+        break
+      } else {
+        logger.debug('waiting for an account to become available...')
+      }
+    } while (await tick())
+
+    return resolvedAccount
+  }
+
+  /**
+   * Sign a transaction using a specific child account
+   * @param address {string} - The address of the account to sign the tx
+   * @param tx {object} - The transaction object
+   * @returns {object} The signed transaction object from web3.js
+   */
+  async signTx(address, tx) {
+    this._enforceChild(address)
+    await this.init()
+
+    const privKey = this.accounts[address].wallet.getPrivateKey()
+    return await this.web3.eth.accounts.signTransaction(
+      tx,
+      bufferToHex(privKey)
+    )
+  }
+
+  /**
+   * Send a transaction from an available sender account.
+   *
+   * NOTE: This does not wait for the transaction to be mined!
+   *
+   * @param tx {object} - The transaction object, sans `from` and `nonce`
+   * @param onReceipt {function} - An optional callback to call when a receipt is found
+   * @returns {string} The transaction hash of the sent transaction
+   */
+  async sendTx(tx, onReceipt) {
+    const address = await this.getAvailableAccount()
+
+    // Set the from and nonce for the account
+    tx = {
+      ...tx,
+      from: address,
+      nonce: await this.txCount(address)
+    }
+
+    if (!tx.gasPrice) {
+      tx.gasPrice = await this.gasPrice()
+    }
+
+    logger.debug('Sending tx: ', tx)
+
+    const signed = await this.signTx(address, tx)
+    const rawTx = signed.rawTransaction
+    // Not in this version of web3.js 1.0...
+    // const txHash = signed.transactionHash
+    const txHash = this.web3.utils.sha3(rawTx)
+
+    // If there's an onReceipt cb, store it for later
+    if (onReceipt) {
+      if (typeof onReceipt !== 'function') {
+        throw new Error('onReceipt is not a function')
+      }
+      this.receiptCallbacks[txHash] = onReceipt
+    }
+
+    // In case it needs to be rebroadcast
+    this.pendingTransactions[txHash] = rawTx
+
+    // blast it
+    await sendRawTransaction(this.web3, rawTx)
+
+    await this.incrementTxCount(address)
```

Comment:
I don't see anywhere that we decrement txCount. What happens if the transaction fails ? For example if we have a bug and submit an invalid tx ? Wouldn't that cause subsequent transactions to get stuck because there would be a hole in the nonce sequence ?

I wonder if it's worth subscribing to errors and decrementing the nonce... ?

Reply:
Fairly certain a failed tx is still counted in the nonce as it's still included in a block.  Dropped transactions however, are a little bit of a concern here as I've noted in a couple places in the source.  The re-broadcasting should in the best of times make this a non-issue.  But if there's a pending tx, the service is restarted(lost from memory), and a tx is dropped, that will cause a problem.

I'm just not sure if that edge case needs to be dealt with now or later before volume increases.  I'd probably defer to your opinion on that.

Reply:
My latest commented change should take care of this edge case I mentioned.  I think this can be resolved?